src/Cli/dotnet/commands/dotnet-workload/install/NetSdkMsiInstallerClient.cs

```diff
@@ -108,38 +128,64 @@ public void GarbageCollect(Func<string, IWorkloadResolver> getResolverForWorkloa
                 Log?.LogMessage($"Starting garbage collection.");
                 Log?.LogMessage($"Garbage Collection Mode: CleanAllPacks={cleanAllPacks}.");
 
-                IEnumerable<SdkFeatureBand> installedFeatureBands = GetInstalledFeatureBands(Log);
-                IEnumerable<WorkloadId> installedWorkloads = RecordRepository.GetInstalledWorkloads(_sdkFeatureBand);
+                var garbageCollector = new WorkloadGarbageCollector(DotNetHome, _sdkFeatureBand, RecordRepository.GetInstalledWorkloads(_sdkFeatureBand), getResolverForWorkloadSet, new SetupLogReporter(Log));
+                garbageCollector.Collect();
 
-                var installedPacks = installedWorkloads.SelectMany(workload => _workloadResolver.GetPacksInWorkload(workload))
-                    .Distinct();
 
-                var installedPackInfos = installedPacks.Select(pack => _workloadResolver.TryGetPackInfo(pack))
-                    .Where(pack => pack != null);
+                IEnumerable<SdkFeatureBand> installedFeatureBands = GetInstalledFeatureBands(Log);
 
-                //  The same workload pack may be aliased from two different names, for example
-                //  Microsoft.NETCore.App.Runtime.Mono.android-arm is aliased from Microsoft.NETCore.App.Runtime.Mono.net6.android-arm and
-                //  from Microsoft.NETCore.App.Runtime.Mono.net6.android-arm64
-                HashSet<(WorkloadPackId id, string version)> expectedWorkloadPacks = new HashSet<(WorkloadPackId id, string version)>();
-                foreach (var expectedPack in installedPackInfos)
+                List<WorkloadManifestRecord> manifestsToRemove = new();
+
+                var installedWorkloadManifests = GetWorkloadManifestRecords();
+                foreach (var manifestRecord in installedWorkloadManifests)
                 {
-                    if (!expectedPack.Id.ToString().Equals(expectedPack.ResolvedPackageId, StringComparison.OrdinalIgnoreCase))
+                    DependencyProvider depProvider = new DependencyProvider(manifestRecord.ProviderKeyName);
+
+                    (bool shouldBeInstalled, string reason) shouldBeInstalled(SdkFeatureBand dependentFeatureBand)
```

Comment:
It seems that we (were) recommending using PascalCase for local functions. https://stackoverflow.com/a/75593010/294804

Not sure that is still the case but I've generally treated them as methods, so I've used PascalCase myself.